File: controls/progressdialog.cpp
#include "progressdialog.h"

namespace NickvisionTubeConverter::Controls
{
    ProgressDialog::ProgressDialog(Gtk::Window& parent, const std::string& description, const std::function<void()>& work) : Gtk::Dialog("Please Wait", parent, true, true), m_work(work), m_isFinished(false)
    {
        //==Settings==//
        set_default_size(400, 120);
        set_resizable(false);
        set_hide_on_close(true);
        set_deletable(false);
        m_connectionTimeout = Glib::signal_timeout().connect(sigc::mem_fun(*this, &ProgressDialog::timeout), 50);
        //==Description==//
        m_lblDescription.set_markup("<b>" + description + "</b>");
        m_lblDescription.set_halign(Gtk::Align::START);
        m_lblDescription.set_margin(6);
        //==Progress Bar==//
        m_progBar.set_margin(10);
        //==Layout==//
        m_mainBox.set_orientation(Gtk::Orientation::VERTICAL);
        m_mainBox.append(m_lblDescription);
        m_mainBox.append(m_progBar);
        set_child(m_mainBox);
        //==Thread==//
        m_thread = std::jthread([&]()
        {
            m_work();
            m_isFinished = true;
        });
    }

    ProgressDialog::~ProgressDialog()
    {
        m_connectionTimeout.disconnect();
    }

    void ProgressDialog::show()
    {
        if(!m_isFinished)
        {
            Gtk::Dialog::show();
        }
    }

    bool ProgressDialog::timeout()
    {
        m_progBar.pulse();
        if(m_isFinished)
        {
            hide();
        }
        return true;
    }
}

File: NickvisionTubeConverter.GNOME/Views/ProgressDialog.cs
using NickvisionTubeConverter.Shared.Controllers;
using NickvisionTubeConverter.Shared.Models;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace NickvisionTubeConverter.GNOME.Views;

public class ProgressDialog
{
    private readonly Adw.Window _window;
    private readonly Gtk.Label _lblDescription;
    private readonly Gtk.ProgressBar _progBar;
    private readonly Gtk.Box _mainBox;

    public ProgressDialog(Gtk.Window parent, string description)
    {
        _window = Adw.Window.New();
        //Window Settings
        _window.SetTransientFor(parent);
        _window.SetDefaultSize(400, 60);
        _window.SetModal(true);
        _window.SetResizable(false);
        _window.SetDeletable(false);
        _window.SetDestroyWithParent(false);
        //Description Label
        _lblDescription = Gtk.Label.New(null);
        _lblDescription.SetMarkup("<b>" + description + "</b>");
        _lblDescription.SetHalign(Gtk.Align.Start);
        //Progress Bar
        _progBar = Gtk.ProgressBar.New();
        //Main Box
        _mainBox = Gtk.Box.New(Gtk.Orientation.Vertical, 20);
        _mainBox.SetMarginStart(10);
        _mainBox.SetMarginTop(10);
        _mainBox.SetMarginEnd(10);
        _mainBox.SetMarginBottom(10);
        _mainBox.Append(_lblDescription);
        _mainBox.Append(_progBar);
        _window.SetContent(_mainBox);
    }
}